### DISCUSSION TITLE
Can I2S be used so the RP2040 is slave?

### DISCUSSION THREAD

**Discussion by frohro** [earlephilhower/arduino-pico#1897] [Q&A] [answered]
With the present I2S implementation, I get a sample rate of 48.047 kHz when I want 48 kHz.  I have a clock that is accurate to within 10ppm at 12.288 MHz (100x more accurate and 256 times 48 kHz) easily available and I am planning to send this I2S data to another computer where it will be processed and output with a 48 kHz sample rate in real-time on the second computer's sound card.   I am worried that the mismatch in sample rates will mess up my DSP.  I know I can discard samples or some other fancier ways of trying to match sample rates, but it would be much better if I didn't have to as often.  If it was possible to use my accurate clock and let the Pico be in I2S slave mode as opposed to master, I think this would be better for my application.  The ADC I'm using is the [PCM1808](https://www.ti.com/lit/ds/symlink/pcm1808.pdf?ts=1702091774603&ref_url=https%253A%252F%252Fwww.google.com%252F).  Can we use the Pico in slave mode with I2S?

**Comment by earlephilhower** [comment, marked as answer]
Nope, no slave support.  There is, sadly, no HW support for clocking the PIOs directly from external pins.  You could theoretically rewrite the I2S.pio code to do so by polling and looping on the clock pin for every bit in or out, but you would need to be very careful about setup and hold timing (and making sure it still fits in the PIO memory, of course!) vs. the internal PIO clock frequency.

You might be able to clock the whole Pico off of your 12.88MHz source, though, (through its pll), and making sure you have integer dividers to keep things nice and neat.   That's a HW mod and not something I could really be of much help with.

**Reply by frohro** [reply]
Thanks @earlephilhower for the helpful answer.  I guess I have to accept 48.047 kHz sampling rate.  If I change the crystal, then the USB frequency is off by significantly more than the ADC frequency which is likely to be a problem.  (4*12.288 = 49.152)
Sad, but thanks for the help.  The slave mode idea will not work.

**Comment by frohro** [comment]
Hi @earlephilhower,
The lack of a slave mode for the Pico is a pretty severe limitation if you want to have a high-performance ADC run on a low-jitter clock and still run at standard sample rates like 96 kHz.  It seems you could make a slave mode by using the WAIT of the PIO.  You have at least four clock cycles, even if your SKCI is running at 24.576 MHz (96 kHz * 256).  Other folk have accomplished this, and there is code out there that might help, for example: https://github.com/malacalypse/rp2040_i2s_example
With slave ability, we could use a low-jitter external clock with the standard Raspberry Pi Pico and have a high-quality audio ADC.  I vote for this feature, so the Pico can catch up to ESP32 and SAMD devices.  :-)

**Reply by earlephilhower** [reply]
An I2S input peripheral would have the PIO interface and the DMA back end.  Looks like you have both, and some hardware to develop for.  Why not try making a PR for it? I myself don't have any I2S master ADC boards in my junk pile.

ADCInput shows how to use the ABM to handle DMA input into the standard Stream API returning int16_t on a read.

**Comment by frohro** [comment]
Thanks for the vote of confidence, @earlephilhower.  I might be able to modify one of my boards to make the PCM1808 master, but without that, I don't have anything to test with either.  :-)  I specifically used your advice not to make them that way, even though there were advantages.  :-)  Perhaps the biggest one for me would be matching the sample rate to 48 kHz.  I think I was able to do that, maybe well enough as mentioned [here](https://github.com/earlephilhower/arduino-pico/discussions/3073).
Thanks,
Rob

**Reply by earlephilhower** [reply]
Ah, too bad.  While it would be possible for me to hack up an I2S sender master from one Pico into a 2nd I2S slave Pico, if both sides were done by me then I know I'd make it work...even if I didn't follow the spec!  :laughing:

Do you know of any boards on AX or similar with I2S master generation?

**Reply by frohro** [reply]
I'm not sure what you mean by "on AX", but you can get the PCM1808 on a breakout board for around $3.  The problem is that you also need to have a clock besides the pico, so you will need a clock board, and you are still doing a good bit of wiring things up.  The next time I do a board spin, I'll see if I can't make it so I can use some jumpers to make my board work either way.  It may be a while, though.  Trump's tariffs have made it more expensive to spin boards all the time.  It is likely that rewiring one I have might happen first.

**Reply by earlephilhower** [reply]
My rewriting of the I2S bidirectional PIO program in #3117 made me think it would be a relatively simple process to do a simple slave output mode as well.  Give #3125 a look.  It makes BCLK and LRCLK into inputs and drives the PIO off of them, sampling LRCLK on BCLK 0->1 and driving data on BCLK 1->0 per the spec AIUI.  I've tested just using an external RP2040 as an I2S clock generator , driving BCLK and LRCLK into the target slave mode, and capturing the combined BCLK/ LRCLK, and DOUT in pulseview.  Looks okay under simple testing...

**Comment by frohro** [comment]
Wow!  Super cool @earlephilhower!  Now I need to hook one of these PCM1808 and an Si5351 as an oscillator and see if I can make a better sound card.  School is starting next week, so it may take a while for me to test this, but I'm looking forward to it.  Again, thanks for adding slave mode.  It will make making a sound card and an SDR with an external ADC much better.
